The `revive` compiler documentation (#424)
paritytech/revive: book/src/developer_guide/contributing.md

# Contributor guide

The `revive` compiler is an open source software project and we gladly accept quality contributions from anyone!

## Getting started

A quick reference on how to build the Solidity compiler is maintained in the projects [README.md](https://github.com/paritytech/revive?tab=readme-ov-file#building-from-source).

### Using the `Makefile`

The [Makefile](https://github.com/paritytech/revive/blob/main/Makefile) comprehensively encapsulates all development aspects of this codebase. It is kept concise and readable. Please read and use it! You'll learn for example:

- How to build and install a `resolc` development version.
- How to run tests and benchmarks.
- How to cross-compile `resolc`.

As a general rule-of-thumb: If `make test` runs fine locally, chances for green CI pipelines are good.

### Codebase organization

For the most parts, `revive` is a rather standard Rust workspace codebase. There are some non-Rust dependencies, which sometimes complicates things a little bit.

#### The `crates/` dir

> All rust-crates live under the `crates/` directory. The workspace automatically consideres any crate found in there. If you need to add a new create, please implement it there.

Reviewer: ```suggestion
All Rust crates live under the `crates/` directory. The workspace automatically considers any crate found therein. If you need to add a new create, please implement it there.
```

Compiler library crates should be named with the `revive-` prefix. The crate location doesn't need the prefix.

#### Dependencies

> Dependencies should be added as workspace dependencies. Try to avoid pinning dependencies whenever possible. If possible to do so, add dev dependencies as `dev-dependencies` only.

Reviewer: ```suggestion
Dependencies should be added as workspace dependencies. Try to avoid pinning dependencies whenever possible. If possible, add dev dependencies as `dev-dependencies` only.
```

Please do always include the `Cargo.lock` dependency lock file with your PR. Please don't run `cargo update` together with other changes (it is preferred to update the lock file in a dedicated dependency update PR).

## Contribution rules

1. Changes must be submitted via a pull request (PR) to the github upstream repository.
2. Ensure that your branch passes `make test` locally when submitting a pull request.
3. A PR must not be merged until CI fully passes. Exceptions can be made (for example to fix CI issues itself).
4. No force pushes to the `main` branch and open PR branches.
5. Maintainers can request changes or deny contributions to their own discretion.

## Style guide

We require the official Rust formatter and clippy linter. In addition to that, please also consider the following best-effort aspects:

- Avoid [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) and strings. Instead, add them as module constants.
- Avoid abbreviated variable and function names. Always provide meaningful and readable symbols.

> - Don't write macros and don't use thrid party macros for things that can easily be expressed in few lines of code or outlined into functions.

Reviewer: ```suggestion
- Don't write macros and don't use third party macros for things that can easily be expressed in few lines of code or outlined into functions.
```
